rust_cli/src/commands/idl.rs

```diff
@@ -0,0 +1,512 @@
+use anyhow::{Result, anyhow};
+use flate2::write::GzEncoder;
+use flate2::Compression;
+use solana_client::rpc_config::RpcSendTransactionConfig;
+use solana_sdk::commitment_config::CommitmentConfig;
+use solana_sdk::{
+    pubkey::Pubkey,
+    signature::Keypair,
+    transaction::Transaction,
+    signer::Signer,
+    compute_budget::ComputeBudgetInstruction,
+};
+use std::io::Write;
+use std::io::Read;
+use std::{str::FromStr, fs};
+use crate::codama_sdk::instructions::{Resize, ResizeInstructionArgs};
+use crate::config::get_user_config;
+use crate::codama_sdk::{
+    instructions::{
+        initialize::{Initialize, InitializeInstructionArgs},
+        create_buffer::CreateBufferBuilder,
+        write_buffer::{WriteBuffer, WriteBufferInstructionArgs},
+        set_buffer::{SetBuffer, SetBufferInstructionArgs},
+        close_buffer::CloseBuffer,
+    },
+    programs::UPLOAD_IDL_ANCHOR_ID,
+};
+use reqwest::blocking::Client;
+use flate2::read::GzDecoder;
+
+const IDL_SEED: &str = "idl";
+const METADATA_SEED: &str = "metadata";
+const METADATA_OFFSET: usize = 44;
+const CHUNK_SIZE: u16 = 900;
+const MAX_RESIZE_STEP: u16 = 10240;
+
+
+
+pub fn upload_idl_by_json_path(
+    idl_path: &str,
+    program_id: &str,
+    keypair_path: Option<&str>,
+    priority_fees_per_cu: u64,
+) -> Result<()> {
+    upload_data_by_json_path(idl_path, program_id, keypair_path, priority_fees_per_cu, IDL_SEED)
+}
+
+pub fn upload_metadata_by_json_path(
+    metadata_path: &str,
+    program_id: &str,
+    keypair_path: Option<&str>,
+    priority_fees_per_cu: u64,
+) -> Result<()> {
+    upload_data_by_json_path(metadata_path, program_id, keypair_path, priority_fees_per_cu, METADATA_SEED)
+}
+
+pub fn upload_idl_by_json_url(
+    url: &str,
+    program_id: &str,
+    keypair_path: Option<&str>,
+    priority_fees_per_cu: u64,
+) -> Result<()> {
+    let json_data = fetch_json_from_url(url)?;
+    upload_data_from_bytes(json_data, program_id, keypair_path, priority_fees_per_cu, IDL_SEED)
+}
+
+pub fn upload_metadata_by_json_url(
+    url: &str,
+    program_id: &str,
+    keypair_path: Option<&str>,
+    priority_fees_per_cu: u64,
+) -> Result<()> {
+    let json_data = fetch_json_from_url(url)?;
+    upload_data_from_bytes(json_data, program_id, keypair_path, priority_fees_per_cu, METADATA_SEED)
+}
+
+fn fetch_json_from_url(url: &str) -> Result<Vec<u8>> {
+    let client = Client::new();
+    let response = client.get(url)
+        .send()
+        .map_err(|e| anyhow!("Failed to fetch from URL: {}", e))?;
+    
+    if !response.status().is_success() {
+        return Err(anyhow!("Failed to fetch URL: HTTP {}", response.status()));
+    }
+
+    response.bytes()
+        .map_err(|e| anyhow!("Failed to read response body: {}", e))
+        .map(|b| b.to_vec())
+}
+
+// Refactor the existing upload_data_by_json_path to use this new function
```

Comment:
Is this a todo or a left over? upload_data_by_json_path and upload_data_by_json_path are the same name? 🤔